pkg/synthetictests/kubelet.go

```diff
@@ -363,3 +364,70 @@ func testSystemDTimeout(events monitorapi.Intervals) []*ginkgo.JUnitTestCase {
 	// write a passing test to trigger detection of this issue as a flake. Doing this first to try to see how frequent the issue actually is
 	return []*ginkgo.JUnitTestCase{failure, success}
 }
+
+var errImagePullTimeoutRE = regexp.MustCompile("ErrImagePull.*read: connection timed out")
+var errImagePullGenericRE = regexp.MustCompile("ErrImagePull")
+
+func testErrImagePullConnTimeoutOpenShiftNamespaces(events monitorapi.Intervals) []*ginkgo.JUnitTestCase {
+	const testName = "[sig-node] should not encounter ErrImagePull read connection timeout in openshift namespace pods"
+	return buildTestsFailIfRegexMatch(testName, errImagePullTimeoutRE, nil, true, events)
+}
+
+func testErrImagePullConnTimeout(events monitorapi.Intervals) []*ginkgo.JUnitTestCase {
+	const testName = "[sig-node] should not encounter ErrImagePull read connection timeout in non-openshift namespace pods"
+	return buildTestsFailIfRegexMatch(testName, errImagePullTimeoutRE, nil, false, events)
+}
+
+func testErrImagePullGenericOpenShiftNamespaces(events monitorapi.Intervals) []*ginkgo.JUnitTestCase {
+	const testName = "[sig-node] should not encounter ErrImagePull in openshift namespace pods"
+	return buildTestsFailIfRegexMatch(testName, errImagePullGenericRE, errImagePullTimeoutRE, true, events)
+}
+
+func testErrImagePullGeneric(events monitorapi.Intervals) []*ginkgo.JUnitTestCase {
+	const testName = "[sig-node] should not encounter ErrImagePull in non-openshift namespace pods"
+	return buildTestsFailIfRegexMatch(testName, errImagePullGenericRE, errImagePullTimeoutRE, false, events)
+}
+
+func buildTestsFailIfRegexMatch(testName string, matchRE, dontMatchRE *regexp.Regexp, inOCPNS bool, events monitorapi.Intervals) []*ginkgo.JUnitTestCase {
+
+	matchedErrMsgs := []string{}
+	for _, event := range events {
+		estr := event.String()
+
+		// Skip if the namespace doesn't match what the test was interested in:
+		if (inOCPNS && !strings.HasPrefix(estr, "ns/openshift-")) ||
+			(!inOCPNS && strings.HasPrefix(estr, "ns/openshift-")) {
+			continue
+		}
+
+		// Skip if we don't match the search regex:
+		if !matchRE.MatchString(estr) {
+			continue
+		}
+
+		// Skip if we *do* match the don't match regex:
+		if dontMatchRE != nil && dontMatchRE.MatchString(estr) {
+			continue
+		}
+
+		matchedErrMsgs = append(matchedErrMsgs, event.String())
+	}
+
+	success := &ginkgo.JUnitTestCase{Name: testName}
+	if len(matchedErrMsgs) > 0 {
+		return []*ginkgo.JUnitTestCase{success}
+	}
+
+	failure := &ginkgo.JUnitTestCase{
+		Name:      testName,
+		SystemOut: strings.Join(matchedErrMsgs, "\n"),
+		FailureOutput: &ginkgo.FailureOutput{
+			Output: fmt.Sprintf("%d pod event intervals matched regex %s\n\n%s",
```

Comment:
I dislike error messages like this.  Identifying pods and sorting by them or identifying images would be helpful.  the point is not the regex, the point is the impact of the failure.

Reply:
I don't want to get into parsing locators, but I have changed it so the FailureOutput has sorted list of locators (ns/blah pod/foo container/bar) only, and the SystemOut has the full event String. (which  can be very large)